Migrate from core-js@2 to core-js@3 in Babel config

# Feature request

## Is your feature request related to a problem? Please describe.

core-js@3 has been released in March 2019, adding a lot of new features. It's a great time to revamp the Babel config provided by Next.

## Describe the solution you'd like

A modern-ish approach to configure Babel.

## Describe alternatives you've considered

Got rid of `next/babel` preset in custom Babel config, adding everything by hand.

## Additional context

So, I had this quite straightforward Babel config (`.babelrc`) for ages:

```
{
  "plugins": [
    "@babel/plugin-transform-flow-strip-types",
    "babel-plugin-styled-components",
    "inline-react-svg"
  ],
  "presets": [
    "next/babel"
  ]
}
```
with the following `.browserslistrc`:
```
node 10.16
last 4 chrome versions
last 4 firefox versions
last 4 edge versions
last 2 safari versions
ie 11
not dead
```
and eventually I discovered that the app doesn't work in IE 11. I mean, something worked and something didn't. So I started to dig up to make sure every feature which Internet Explorer 11 is missing is polyfilled.

I checked the [next/babel](https://github.com/zeit/next.js/blob/aac5121466f3d58e2c18c4921a9ce36fb7512d2e/packages/next/build/babel/preset.ts) preset and figured out that it contains both `@babel/plugin-transform-runtime` plugin and `@babel/preset-env`:
```
 [
        require('@babel/plugin-transform-runtime'),
        {
          corejs: 2,
          helpers: true,
          regenerator: true,
          useESModules: supportsESM && presetEnvConfig.modules !== 'commonjs',
          ...options['transform-runtime'],
        },
],
```

Here, the polyfilling is being done by the `@babel/plugin-transform-runtime` plugin, using core-js@2. But in the age of Babel ^7.4, it's suggested to use `@babel/preset-env` for that, setting its `useBuiltIns` and `corejs` options.

Being unable to use `@babel/preset-env` and disable `@babel/plugin-transform-runtime`, I "dissected" the whole `next/babel` preset and reassembled it in custom `.babelrc`, excluding some plugins which are irrelevant for my project (like AMP and `styled-jsx`).

So I ended up with this:
```
{
  "plugins": [
    "@babel/plugin-transform-flow-strip-types",
    "inline-react-svg"
  ],
  "env": {
    "development": {
      "plugins": [
        "babel-plugin-styled-components",
        "babel-plugin-react-require",
        "@babel/plugin-syntax-dynamic-import",
        "./node_modules/next/dist/build/babel/plugins/react-loadable-plugin",
        "@babel/plugin-proposal-class-properties",
        [
          "@babel/plugin-proposal-object-rest-spread",
          {
            "useBuiltIns": true
          }
        ]
      ],
      "presets": [
        [
          "@babel/preset-env",
          {
            "debug": true,
            "modules": false,
            "exclude": [
              "transform-typeof-symbol"
            ],
            "useBuiltIns": "usage",
            "corejs": "3.1"
          }
        ],
        [
          "@babel/preset-react",
          {
            "development": true
          }
        ]
      ]
    },
    "production": {
      "plugins": [
        "babel-plugin-styled-components",
        "babel-plugin-react-require",
        "@babel/plugin-syntax-dynamic-import",
        "./node_modules/next/dist/build/babel/plugins/react-loadable-plugin",
        "@babel/plugin-proposal-class-properties",
        [
          "@babel/plugin-proposal-object-rest-spread",
          {
            "useBuiltIns": true
          }
        ]
      ],
      "presets": [
        [
          "@babel/preset-env",
          {
            "modules": false,
            "exclude": [
              "transform-typeof-symbol"
            ],
            "useBuiltIns": "usage",
            "corejs": "3.1"
          }
        ],
        "@babel/preset-react"
      ]
    },
    "test": {
      "plugins": [
        "babel-plugin-styled-components",
        "@babel/plugin-proposal-class-properties"
      ],
      "presets": [
        [
          "@babel/preset-react",
          {
            "development": true
          }
        ],
        [
          "@babel/preset-env",
          {
            "targets": "node 10.16",
            "useBuiltIns": false,
            "ignoreBrowserslistConfig": true
          }
        ]
      ]
    }
  }
}
```
(Note: `babel.config.js` didn't work for me, so I stayed with the `.babelrc` for now)

That being said, please consider upgrading the `next/babel` preset to make sure it's using all the latest/greatest stuff.

Nice project by the way :slightly_smiling_face:.

I've been hit with a similar issue
```
Module not found: Can't resolve 'core-js/library/fn/object/assign' in '../node_modules/@babel/runtime-corejs2/core-js/object'

I think this is because I'm using `core-js@3` elsewhere in my project, but next is expecting v2.
It would be a good time to switch to core-js@3 as there will be no more work on 2, and various projects will start moving to 2

--- a/packages/next/build/babel/preset.ts
+++ b/packages/next/build/babel/preset.ts
@@ -64,9 +64,7 @@ module.exports = (
   const supportsESM = api.caller(supportsStaticESM)
   const isServer = api.caller((caller: any) => !!caller && caller.isServer)
   const isModern = api.caller((caller: any) => !!caller && caller.isModern)
-  const isPolyfillsOptimization = api.caller(
-    (caller: any) => !!caller && caller.polyfillsOptimization
-  )
+
   const isLaxModern =
     isModern ||
     (options['preset-env']?.targets &&
@@ -155,7 +153,7 @@ module.exports = (
       !isServer && [
         require('@babel/plugin-transform-runtime'),
         {
-          corejs: isPolyfillsOptimization ? false : 2,
+          corejs: false,
           helpers: true,
           regenerator: true,
           useESModules: supportsESM && presetEnvConfig.modules !== 'commonjs',

--- a/packages/next/build/webpack-config.ts
+++ b/packages/next/build/webpack-config.ts
@@ -54,10 +54,7 @@ const escapePathVariables = (value: any) => {
     : value
 }
 
-function getOptimizedAliases(
-  isServer: boolean,
-  polyfillsOptimization: boolean
-): { [pkg: string]: string } {
+function getOptimizedAliases(isServer: boolean): { [pkg: string]: string } {
   if (isServer) {
     return {}
   }
@@ -68,12 +65,6 @@ function getOptimizedAliases(
   const shimAssign = path.join(__dirname, 'polyfills', 'object.assign')
   return Object.assign(
     {},
-    // Polyfill: Window#fetch
-    polyfillsOptimization
-      ? undefined
-      : {
-          __next_polyfill__fetch: require.resolve('whatwg-fetch'),
-        },
     {
       unfetch$: stubWindowFetch,
       'isomorphic-unfetch$': stubWindowFetch,
@@ -84,13 +75,6 @@ function getOptimizedAliases(
         'whatwg-fetch.js'
       ),
     },
-    polyfillsOptimization
-      ? undefined
-      : {
-          // Polyfill: Object.assign
-          __next_polyfill__object_assign: require.resolve('object-assign'),
-          '@babel/runtime-corejs2/core-js/object/assign': stubObjectAssign,
-        },
     {
       'object-assign$': stubObjectAssign,
 
@@ -168,7 +152,6 @@ export default async function getBaseWebpackConfig(
         babelPresetPlugins,
         hasModern: !!config.experimental.modern,
         development: dev,
-        polyfillsOptimization: !!config.experimental.polyfillsOptimization,
       },
     },
     // Backwards compat
@@ -215,9 +198,7 @@ export default async function getBaseWebpackConfig(
           ),
         [CLIENT_STATIC_FILES_RUNTIME_POLYFILLS]: path.join(
           NEXT_PROJECT_ROOT_DIST_CLIENT,
-          config.experimental.polyfillsOptimization
-            ? 'polyfills-nomodule.js'
-            : 'polyfills.js'
+          'polyfills.js'
         ),
       } as ClientEntries)
     : undefined
@@ -267,17 +248,7 @@ export default async function getBaseWebpackConfig(
       next: NEXT_PROJECT_ROOT,
       [PAGES_DIR_ALIAS]: pagesDir,
       [DOT_NEXT_ALIAS]: distDir,
-      ...getOptimizedAliases(
-        isServer,
-        !!config.experimental.polyfillsOptimization
-      ),
-
-      // Temporary to allow runtime-corejs2 to be stubbed in experimental polyfillsOptimization
-      ...(config.experimental.polyfillsOptimization
-        ? {
-            '@babel/runtime-corejs2': '@babel/runtime',
-          }
-        : undefined),
+      ...getOptimizedAliases(isServer),
     },
     mainFields: isServer ? ['main', 'module'] : ['browser', 'module', 'main'],
     plugins: [PnpWebpackPlugin],
@@ -528,11 +499,7 @@ export default async function getBaseWebpackConfig(
               !res.match(/next[/\\]dist[/\\]next-server[/\\]/) &&
               (res.match(/[/\\]next[/\\]dist[/\\]/) ||
                 // This is the @babel/plugin-transform-runtime "helpers: true" option
-                res.match(/node_modules[/\\]@babel[/\\]runtime[/\\]/) ||
-                (!config.experimental.polyfillsOptimization &&
-                  res.match(
-                    /node_modules[/\\]@babel[/\\]runtime-corejs2[/\\]/
-                  )))
+                res.match(/node_modules[/\\]@babel[/\\]runtime[/\\]/))
             ) {
               return callback()
             }
@@ -727,9 +694,6 @@ export default async function getBaseWebpackConfig(
         ),
         'process.env.__NEXT_MODERN_BUILD': JSON.stringify(
           config.experimental.modern && !dev
-        ),
-        'process.env.__NEXT_POLYFILLS_OPTIMIZATION': JSON.stringify(
-          !!config.experimental.polyfillsOptimization
         ),
         'process.env.__NEXT_GRANULAR_CHUNKS': JSON.stringify(
           config.experimental.granularChunks && !dev

--- a/packages/next/build/webpack/loaders/next-babel-loader.js
+++ b/packages/next/build/webpack/loaders/next-babel-loader.js
@@ -59,7 +59,6 @@ module.exports = babelLoader.custom(babel => {
         hasModern: opts.hasModern,
         babelPresetPlugins: opts.babelPresetPlugins,
         development: opts.development,
-        polyfillsOptimization: opts.polyfillsOptimization,
       }
       const filename = join(opts.cwd, 'noop.js')
       const loader = Object.assign(
@@ -72,7 +71,7 @@ module.exports = babelLoader.custom(babel => {
                 (opts.isServer ? '-server' : '') +
                 (opts.isModern ? '-modern' : '') +
                 (opts.hasModern ? '-has-modern' : '') +
-                (opts.polyfillsOptimization ? '-new-polyfills' : '') +
+                '-new-polyfills' +
                 (opts.development ? '-development' : '-production') +
                 JSON.stringify(
                   babel.loadPartialConfig({
@@ -95,7 +94,6 @@ module.exports = babelLoader.custom(babel => {
       delete loader.hasModern
       delete loader.pagesDir
       delete loader.babelPresetPlugins
-      delete loader.polyfillsOptimization
       delete loader.development
       return { loader, custom }
     },
@@ -110,7 +108,6 @@ module.exports = babelLoader.custom(babel => {
           pagesDir,
           babelPresetPlugins,
           development,
-          polyfillsOptimization,
         },
       }
     ) {
@@ -134,7 +131,6 @@ module.exports = babelLoader.custom(babel => {
 
       options.caller.isServer = isServer
       options.caller.isModern = isModern
-      options.caller.polyfillsOptimization = polyfillsOptimization
       options.caller.isDev = development
 
       options.plugins = options.plugins || []

--- a/packages/next/client/index.js
+++ b/packages/next/client/index.js
@@ -14,20 +14,9 @@ import { isDynamicRoute } from '../next-server/lib/router/utils/is-dynamic'
 
 /// <reference types="react-dom/experimental" />
 
-if (process.env.__NEXT_POLYFILLS_OPTIMIZATION) {
-  if (!('finally' in Promise.prototype)) {
-    // eslint-disable-next-line no-extend-native
-    Promise.prototype.finally = require('finally-polyfill')
-  }
-} else {
-  // Polyfill Promise globally
-  // This is needed because Webpack's dynamic loading(common chunks) code
-  // depends on Promise.
-  // So, we need to polyfill it.
-  // See: https://webpack.js.org/guides/code-splitting/#dynamic-imports
-  if (!self.Promise) {
-    self.Promise = require('@babel/runtime-corejs2/core-js/promise')
-  }
+if (!('finally' in Promise.prototype)) {
+  // eslint-disable-next-line no-extend-native
+  Promise.prototype.finally = require('finally-polyfill')
 }
 
 const data = JSON.parse(document.getElementById('__NEXT_DATA__').textContent)

--- a/packages/next/client/polyfills-nomodule.js
+++ b/packages/next/client/polyfills-nomodule.js
@@ -1,1 +1,0 @@
-import '@next/polyfill-nomodule'

--- a/packages/next/client/polyfills.js
+++ b/packages/next/client/polyfills.js
@@ -1,3 +1,1 @@
-import '__next_polyfill__fetch'
-import 'url-polyfill'
-Object.assign = require('__next_polyfill__object_assign')
+import '@next/polyfill-nomodule'

--- a/packages/next/package.json
+++ b/packages/next/package.json
@@ -72,7 +72,6 @@
     "@babel/preset-react": "7.7.0",
     "@babel/preset-typescript": "7.7.2",
     "@babel/runtime": "7.7.2",
-    "@babel/runtime-corejs2": "7.7.2",
     "@babel/types": "7.7.4",
     "@next/polyfill-nomodule": "9.2.2",
     "amphtml-validator": "1.0.23",
@@ -117,7 +116,6 @@
     "mini-css-extract-plugin": "0.8.0",
     "mkdirp": "0.5.1",
     "node-fetch": "2.6.0",
-    "object-assign": "4.1.1",
     "ora": "3.4.0",
     "path-to-regexp": "6.1.0",
     "pnp-webpack-plugin": "1.5.0",
@@ -148,8 +146,7 @@
     "webpack": "4.41.2",
     "webpack-dev-middleware": "3.7.0",
     "webpack-hot-middleware": "2.25.0",
-    "webpack-sources": "1.4.3",
-    "whatwg-fetch": "3.0.0"
+    "webpack-sources": "1.4.3"
   },
   "peerDependencies": {
     "react": "^16.6.0",

--- a/packages/next/taskfile-babel.js
+++ b/packages/next/taskfile-babel.js
@@ -95,11 +95,6 @@ module.exports = function(task) {
     const output = transform(file.data, options)
     const ext = extname(file.base)
 
-    output.code = output.code.replace(
-      /@babel\/runtime\//g,
-      '@babel/runtime-corejs2/'
-    )
-
     // Replace `.ts|.tsx` with `.js` in files with an extension
     if (ext) {
       const extRegex = new RegExp(ext.replace('.', '\\.') + '$', 'i')

--- a/test/integration/bundle-size-profiling/next.config.js
+++ b/test/integration/bundle-size-profiling/next.config.js
@@ -1,7 +1,4 @@
 module.exports = {
-  experimental: {
-    polyfillsOptimization: true,
-  },
   webpack(config, options) {
     if (!options.isServer) {
       config.profile = true

--- a/test/integration/modern-mode/test/index.test.js
+++ b/test/integration/modern-mode/test/index.test.js
@@ -40,14 +40,7 @@ describe('Modern Mode', () => {
   it('should generate client side modern and legacy build files', async () => {
     const buildId = readFileSync(join(appDir, '.next/BUILD_ID'), 'utf8')
 
-    const expectedFiles = [
-      'index',
-      '_app',
-      '_error',
-      'main',
-      'webpack',
-      'commons',
-    ]
+    const expectedFiles = ['index', '_app', '_error', 'main', 'webpack']
     const buildFiles = [
       ...readdirSync(join(appDir, '.next/static', buildId, 'pages')),
       ...readdirSync(join(appDir, '.next/static/runtime')).map(

--- a/test/integration/polyfilling-minimal/next.config.js
+++ b/test/integration/polyfilling-minimal/next.config.js
@@ -1,5 +1,1 @@
-module.exports = {
-  experimental: {
-    polyfillsOptimization: true,
-  },
-}
+module.exports = {}

--- a/test/integration/size-limit/test/index.test.js
+++ b/test/integration/size-limit/test/index.test.js
@@ -80,7 +80,7 @@ describe('Production response size', () => {
     )
 
     // These numbers are without gzip compression!
-    const delta = responseSizesBytes - 233 * 1024
+    const delta = responseSizesBytes - 241 * 1024
     expect(delta).toBeLessThanOrEqual(1024) // don't increase size more than 1kb
     expect(delta).toBeGreaterThanOrEqual(-1024) // don't decrease size more than 1kb without updating target
   })
@@ -100,7 +100,7 @@ describe('Production response size', () => {
     )
 
     // These numbers are without gzip compression!
-    const delta = responseSizesBytes - 201 * 1024
+    const delta = responseSizesBytes - 171 * 1024
     expect(delta).toBeLessThanOrEqual(1024) // don't increase size more than 1kb
     expect(delta).toBeGreaterThanOrEqual(-1024) // don't decrease size more than 1kb without updating target
   })

--- a/test/unit/next-babel.test.js
+++ b/test/unit/next-babel.test.js
@@ -69,7 +69,7 @@ describe('next/babel', () => {
       expect(output).toMatch(`__jsx("a",{href:"/"`)
 
       expect(babel(`const a = ()=><a href="/">home</a>`)).toMatchInlineSnapshot(
-        `"\\"use strict\\";var _interopRequireDefault=require(\\"@babel/runtime-corejs2/helpers/interopRequireDefault\\");var _react=_interopRequireDefault(require(\\"react\\"));var __jsx=_react[\\"default\\"].createElement;var a=function a(){return __jsx(\\"a\\",{href:\\"/\\"},\\"home\\");};"`
+        `"\\"use strict\\";var _interopRequireDefault=require(\\"@babel/runtime/helpers/interopRequireDefault\\");var _react=_interopRequireDefault(require(\\"react\\"));var __jsx=_react[\\"default\\"].createElement;var a=function a(){return __jsx(\\"a\\",{href:\\"/\\"},\\"home\\");};"`
       )
     })
 

